updater.js

```diff
@@ -0,0 +1,225 @@
+  if (!packagePath) {
+    throw new BrowserslistError(
+      'Cannot find package.json. ' +
+      'Is it a right project to run npx browserslist --update-db?'
+    )
+  }
+
+  var rootDir = path.dirname(packagePath)
+
+  var packageManager = ''
+  var lockfile = ''
+
+  var lockfileNpm = path.join(rootDir, 'package-lock.json')
+  var lockfileYarn = path.join(rootDir, 'yarn.lock')
+  var lockfilePnpm = path.join(rootDir, 'pnpm-lock.yaml')
+
+  if (fs.existsSync(lockfileNpm)) {
+    packageManager = 'npm'
+    lockfile = lockfileNpm
+  } else if (fs.existsSync(lockfileYarn)) {
+    packageManager = 'yarn'
+    lockfile = lockfileYarn
+  } else if (fs.existsSync(lockfilePnpm)) {
+    packageManager = 'pnpm'
+    lockfile = lockfilePnpm
+  } else {
+    throw new BrowserslistError('Lockfile search error')
+  }
+
+  return {
+    packageManager: packageManager,
+    lockfile: lockfile
+  }
+}
+
+function getCurrentVersion (lockfileRaw, packageManager) {
+  if (packageManager === 'npm') {
+    var parsedFile = JSON.parse(lockfileRaw)
+    if (
+      parsedFile.dependencies !== undefined &&
+      parsedFile.dependencies[PACKAGE_CANIUSE] !== undefined &&
+      parsedFile.dependencies[PACKAGE_CANIUSE].version !== undefined
+    ) {
+      return parsedFile.dependencies[PACKAGE_CANIUSE].version
+    }
+    return null
+  }
+
+  var parsedLockfileLines = lockfileRaw.split('\n')
+  var countLines = parsedLockfileLines.length
+  if (packageManager === 'yarn') {
+    var foundCaniuse = false
+    for (var i = 0; i < countLines; i++) {
+      var row = parsedLockfileLines[i]
+      if (row.indexOf(PACKAGE_CANIUSE) === 0) {
+        foundCaniuse = true
+        continue
+      }
+
+      if (foundCaniuse && /version/.test(row)) {
+        var rowVersion = /version "([^"]+)"/.exec(row)
+        if (rowVersion.length > 1) {
+          return rowVersion[1]
+        }
+        return null
+      }
+    }
+  } else if (packageManager === 'pnpm') {
+    for (var j = 0; j < countLines; j++) {
+      var line = parsedLockfileLines[j]
+      if (line.indexOf('/' + PACKAGE_CANIUSE) >= 0) {
+        var rule = /\/([\d.]+):/.exec(line)
+        if (rule.length > 1) {
+          return rule[1]
+        }
+        return null
+      }
+    }
+  }
+
+  return null
+}
+
+function getLastVersionInfo () {
+  try {
+    var raw = childProcess
+      .execSync('npm show ' + PACKAGE_CANIUSE + ' --json')
+      .toString()
+
+    return JSON.parse(raw)
+  } catch (e) {
+    throw new BrowserslistError(
+      'An error occurred getting information about the latest version of ' +
+      PACKAGE_CANIUSE +
+      '.\n' +
+      'Check your Internet connection.'
+    )
+  }
+}
+
+function updateLockfile (lockfileRaw, info, packageManager) {
+  var parsedLockfileLines = lockfileRaw.split('\n')
+  var countLines = parsedLockfileLines.length
+  var hasNecessaryDeps = false
+
+  if (packageManager === 'npm') {
+    for (var i = 0; i < parsedLockfileLines.length; i++) {
```

Comment:
Should be work with `package-lock.json` as with JSON? It will be simpler and safer code.

Reply:
I agree that the work will be safer, but the search for nested dependencies will be complicated by tree smash. I try that improvement.

Reply:
You can try recursion to make it easy:

```js
function process (node) {
  if (check(node) {
    node = changeNode(node)
  }
  node.children = node.children.map(i => process(i))
  return node
}
```